common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -3040,6 +3041,50 @@ static void NativeCrypto_ED25519_keypair(JNIEnv* env, jclass, jbyteArray outPubl
     JNI_TRACE("ED25519_keypair(%p, %p) => success", outPublicArray, outPrivateArray);
 }
 
+static jbyteArray NativeCrypto_XWING_public_key_from_seed(JNIEnv* env, jclass,
+                                                          jbyteArray privateKeySeed) {
+    CHECK_ERROR_QUEUE_ON_RETURN;
+
+    ScopedByteArrayRO seedArray(env, privateKeySeed);
+    if (seedArray.get() == nullptr) {
+        JNI_TRACE("NativeCrypto_XWING_public_key_from_seed => privateKeySeed == null");
+        return nullptr;
+    }
+
+    if (seedArray.size() != XWING_PRIVATE_KEY_BYTES) {
+        conscrypt::jniutil::throwException(env, "java/lang/IllegalArgumentException",
+                                           "privateKeySeed length != 32");
+        return nullptr;
+    }
+
+    CBS cbs;
+    CBS_init(&cbs, reinterpret_cast<const uint8_t*>(seedArray.get()), seedArray.size());
+    XWING_private_key privateKey;
+    if (!XWING_parse_private_key(&privateKey, &cbs)) {
+        JNI_TRACE("XWING_parse_private_key failed");
+        conscrypt::jniutil::throwIllegalArgumentException(env, "XWING_parse_private_key failed");
+        return nullptr;
+    }
+
+    ScopedLocalRef<jbyteArray> publicKeyRef(
+            env, env->NewByteArray(static_cast<jsize>(XWING_PUBLIC_KEY_BYTES)));
+    if (publicKeyRef.get() == nullptr) {
+        return nullptr;
+    }
+    ScopedByteArrayRW publicKeyArray(env, publicKeyRef.get());
```

Comment:
This is actually fairly subtle and I wonder if we should make this clearer.

The way this works is that the byte array is written back at the point the destructor is called. Maybe it would be clearer if the lifetime is made more explicit with a scoped block...